### PR TITLE
[6X] Fix issue caused by primary restart after streamed an incomplete cross-segment WAL

### PR DESCRIPTION
**Pull request by huansong**
Backport from upstream commit 148c6ee3be39be61968d580eaafc93795aed1f0c
The corresponding 7X PR for GPDB is https://github.com/greenplum-db/gpdb/pull/13648

GPDB related message
====================
We saw similar issue in GPDB where an incomplete cross-segment WAL
record can cause the mirror to try to stream from the next segment,
whereas the primary has a flush point somewhere at the end of the
previous segment. So the primary saw:
    requested starting point D81/54000000 is ahead of the WAL flush position of this server D81/53FF7FE8

In fact, the test scenario in this commit (026_overwrite_contrecord.pl)
would produce the above error w/o the fix, instead of the one mentioned
in the upstream commit message, at least for GPDB.
But the main difference is just whether the primary has written &
flushed enough data to move its flush point beyond the segment boundary.
The fix would fix both errors.

Note that, because there are some differences between GPDB6 (which
based on PG9.4) and PG9.6 which is the closest version that this fix
has been backported to in upstream, we had to make some changes in
order to cherrypick it to GPDB6:
* Move VerifyOverwriteContrecord() to StartupXLOG because we need the
  xlogreader which isn't there in xlog_redo.
* Change CreateOverwriteContrecordRecord() accordingly because of the
  difference in XLogInsert();
* Disable log collector because we need to examine the log. I brought
  the related changes from GPDB7 for this.
* Use AO table insertion instead of pg_logical_emit_message() which
  isn't there in GPDB6.
* Disable hot standby in the test because it causes an issue. Have to
  use sleep instead of querying the mirror to wait for it to catch up.

Original commit message
==========================

Physical replication always ships WAL segment files to replicas once
they are complete.  This is a problem if one WAL record is split across
a segment boundary and the primary server crashes before writing down
the segment with the next portion of the WAL record: WAL writing after
crash recovery would happily resume at the point where the broken record
started, overwriting that record ... but any standby or backup may have
already received a copy of that segment, and they are not rewinding.
This causes standbys to stop following the primary after the latter
crashes:
  LOG:  invalid contrecord length 7262 at A8/D9FFFBC8
because the standby is still trying to read the continuation record
(contrecord) for the original long WAL record, but it is not there and
it will never be.  A workaround is to stop the replica, delete the WAL
file, and restart it -- at which point a fresh copy is brought over from
the primary.  But that's pretty labor intensive, and I bet many users
would just give up and re-clone the standby instead.

A fix for this problem was already attempted in commit 515e3d84a0b5, but
it only addressed the case for the scenario of WAL archiving, so
streaming replication would still be a problem (as well as other things
such as taking a filesystem-level backup while the server is down after
having crashed), and it had performance scalability problems too; so it
had to be reverted.

This commit fixes the problem using an approach suggested by Andres
Freund, whereby the initial portion(s) of the split-up WAL record are
kept, and a special type of WAL record is written where the contrecord
was lost, so that WAL replay in the replica knows to skip the broken
parts.  With this approach, we can continue to stream/archive segment
files as soon as they are complete, and replay of the broken records
will proceed across the crash point without a hitch.

Because a new type of WAL record is added, users should be careful to
upgrade standbys first, primaries later. Otherwise they risk the standby
being unable to start if the primary happens to write such a record.

A new TAP test that exercises this is added, but the portability of it
is yet to be seen.

This has been wrong since the introduction of physical replication, so
backpatch all the way back.  In stable branches, keep the new
XLogReaderState members at the end of the struct, to avoid an ABI
break.

Author: Álvaro Herrera <alvherre@alvh.no-ip.org>
Reviewed-by: Kyotaro Horiguchi <horikyota.ntt@gmail.com>
Reviewed-by: Nathan Bossart <bossartn@amazon.com>
Discussion: https://postgr.es/m/202108232252.dh7uxf6oxwcy@alvherre.pgsql

### CODE REVIEW

File: src/test/perl/PostgresNode.pm [hunk 1]
```diff
@@ -424,6 +424,12 @@ sub init
 	# print $conf "wal_retrieve_retry_interval = '500ms'\n";
 	print $conf "port = $port\n";
 
+	# In GPBB, logging_collector is enabled by default. Disable it, so that
```
Comment (soumyadeep2007) [inline src/test/perl/PostgresNode.pm:427]: s/In GPBB/GPDB:

File: src/test/recovery/t/026_overwrite_contrecord.pl [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+note "start ",
+  $node->safe_psql('postgres', 'select pg_current_xlog_insert_location()');
+my $initfile = $node->safe_psql('postgres',
+	'SELECT pg_xlogfile_name(pg_current_xlog_insert_location())');
+$node->safe_psql('postgres',
+qq{INSERT INTO t SELECT 1, array_agg(x) from generate_series(1, 24000) x}
+);
+
+sleep 1;
```
Comment (soumyadeep2007) [inline src/test/recovery/t/026_overwrite_contrecord.pl:69]: Why is this sleep necessary?

File: src/test/recovery/t/026_overwrite_contrecord.pl [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+$node_standby->init_from_backup($node, 'backup', has_streaming => 1);
+
+$node_standby->start;
+$node->start;
+
+$node->safe_psql('postgres',
+	qq{create table foo (a text); insert into foo values ('hello')});
+
+# See earlier comments for hot_standby=off. 
+sleep 5;
```
Comment (soumyadeep2007) [inline src/test/recovery/t/026_overwrite_contrecord.pl:98]: Why is this sleep necessary?

**Review by huansong — COMMENTED** [review]
> Can we use cherry-pick -x for the last 2 straight-forward (assuming it was) commits?

Done

File: src/test/recovery/t/026_overwrite_contrecord.pl (hunk 3 above)
Comment (huansong) [reply src/test/recovery/t/026_overwrite_contrecord.pl]: Thanks for the great suggestion!! Yes that works. Done (p.s. it's `replay_location` in 6X).

Comment (soumyadeep2007) [reply src/test/recovery/t/026_overwrite_contrecord.pl]: Super!

**Pull request merged**